vignettes/wip/dm-filtering.Rmd

```diff
@@ -19,18 +19,195 @@ fansi::set_knit_hooks(knitr::knit_hooks)
+dm
+```
+
+Now we know that there are five tables in our `dm` object.
+But how are they connected?
+These relations are best displayed as a visualization of the entity relation model:
+
+```{r}
+cdm_draw(dm)
+```
+
+You can look at a single table with `tbl`.
+Print the `airports` table:
+
+```{r}
+tbl(dm, "airports")
+```
+
+## Filtering a `dm` object {#filter}
+
+`cdm_filter()` allows you to select a subset of a `dm` object.
+
+### How it works
+
+Filtering a `dm` object is not that different from filtering a dataframe or tibble with `dplyr::filter()`.
+
+The corresponding `dm` function is `dm::cdm_filter()` and creates a subset of one table depending on the input.
+But `dm` does not stop here: A cascade of joins is performed along the foreign key relations.
+This results in a new `dm` containing only rows related to the filter operation.
+
+Currently, this only works if the graph induced by the foreign key relations is cycle free.
+Luckily, this is the default for `cdm_nycflights13()`.
+
+### Filtering Examples
+
+Let's see it in action:
+
+**We only want the data that is related with John F. Kennedy International Airport.**
+
+The total number of rows drops from  367 687 to 138 786 (apart from the disconnected `weather` table).
+
+You can see this in the `Rows`-part of the print output.
+
+```{r}
+dm %>% 
+  cdm_filter(airports, name == "John F Kennedy Intl")
+```
+
+Another way to check the number of rows is the function `cdm_nrow()`:
+
+```{r}
+dm %>%
+  cdm_filter(airports, name == "John F Kennedy Intl") %>%
+  cdm_nrow()
+```
+
+Next example:
+
+**Get a `dm` object containing data for flights from New York to the Dulles International Airport in Washington D.C., abbreviated with `IAD`.**
+
+```{r}
+dm %>% 
+  cdm_filter(flights, dest == "IAD")
+```
+
+Of course, you can also chain multiple filters on different tables.
+
+An Example:
+
+**Get all flights from American Airlines which didn't depart from John F. Kennedy International Airport in May 2013.**
+
+```{r}
+dm_aa_may <- dm %>%
+  cdm_filter(airlines, name == "American Airlines Inc.") %>%
+  cdm_filter(airports, name != "John F Kennedy Intl") %>%
+  cdm_filter(flights, month == 5)
+```
+
+You can inspect the filtered tables with `tbl`.
+
+In the table `airlines` American Airlines is the only remaining carrier:
+
+```{r}
+dm_aa_may %>% 
+  tbl("airlines")
+```
+
+Which planes were used to service these flights?
+
+```{r}
+dm_aa_may %>% 
+  tbl("planes")
+```
+
+And indeed, there are only flights from `month==5`:
+
+```{r}
+dm_aa_may %>% 
+  tbl("flights")
+```
+
+For comparison, let's review the equivalent manual query for the `flights` in `dplyr` syntax:
+
+```{r}
+flights %>%
+  left_join(airlines, by = "carrier") %>%
+  filter(name == "American Airlines Inc.") %>%
+  left_join(airports, by = c("origin" = "faa")) %>%
+  filter(
+    name.y != "John F Kennedy Intl",
+    month == 5
+  )
+```
+
+The `dm` code is leaner, because the foreign key information is encoded in the object.
+
```

Comment:
Describe `cdm_apply_filters()` here? What does it do for an in-memory and for a database dm? Earlier, describe the idea of collecting filters before applying them, pointing to this section.